ai_document_search_backend/auth.py

```diff
@@ -13,33 +29,105 @@ class User(BaseModel):
 
 
 class UserInDB(User):
-    password: str
+    hashedPassword: str
+
+
+class Salt(BaseModel):
+    username: str
+    salt: str
+
+
+class SaltInDB(Salt):
+    salt: str
 
 
 fake_users_db = {
     "marius": {
         "username": "marius",
         "full_name": "Marius Berdal Gaalaas",
         "email": "marius@gmail.com",
-        "password": "123",
+        "hashedPassword": "$2b$12$SXT6OzyrQxVZLjW1QAawMe5qUORtJ6RHy1w3cE86U1aV0lc3NDG7S",
+        "disabled": False,
+    },
+    "test": {
+        "username": "test",
+        "full_name": "Test Testesen",
+        "email": "test@gmail.com",
+        "hashedPassword": "$2b$12$jWutkf9LKP9qxQBKwnIjNetU13STTvuZaoBIqmcfCk9pBpZyg1Jey",
         "disabled": False,
-    }
+    },
 }
 
+fake_salts_db = {
+    "marius": {"username": "marius", "salt": "fNRmPKbYOeHEYDor22GcZMN8tSeAT4Ac"},
+    "test": {"username": "test", "salt": "DWs93J1MgEOKQ0Tz0pxgtsnM4I4VNSDz"},
+}
+
+passwordContext = CryptContext(schemes=["bcrypt"], deprecated="auto")
+
+
 router = APIRouter()
 
-oauth2_scheme = OAuth2PasswordBearer(tokenUrl="token")
+oauth2Scheme = OAuth2PasswordBearer(tokenUrl="token")
+
+
+def getHashedPassword(password: str):
+    return passwordContext.hash(password)
+
+
+def checkPassword(password: str, salt: str, hashedPassword: str):
+    saltedPassword = salt + password
+    return passwordContext.verify(saltedPassword, hashedPassword)
+
+
+def createAccessToken(data: dict, expiresDelta: timedelta):
+    dataCopy = data.copy()
+    expireTime = datetime.utcnow() + expiresDelta
+    dataCopy.update({"exp": expireTime})
+    token = jwt.encode(dataCopy, SECRETKEY, algorithm=ALGORITHM)
+    return token
 
 
 # For authenticating
 @router.post("/token")
 async def login(form_data: Annotated[OAuth2PasswordRequestForm, Depends()]):
     user_dict = fake_users_db.get(form_data.username)  # TODO: Make proper users
+    salt = fake_salts_db.get(form_data.username)["salt"]
     if not user_dict:
         raise HTTPException(status_code=400, detail="Incorrect username or password")
     user = UserInDB(**user_dict)
     password = form_data.password
-    if not password == user.password:
+    if not checkPassword(password, salt, user.hashedPassword):
         raise HTTPException(status_code=400, detail="Incorrect username or password")
+    tokenExpires = timedelta(minutes=ACCESS_TOKEN_EXPIRE_MINUTES)
+    token = createAccessToken(data={"sub": user.username}, expiresDelta=tokenExpires)
+
+    return {"access_token": token, "token_type": "bearer"}
+
 
-    return {"access_token": user.username, "token_type": "bearer"}
+# Try to get http://127.0.0.1:8000/auth/main from the fromtend to ensure that tokens are properly implemented
```

Comment:
What is this comment for?

Reply:
The endpoint was meant as a way for the frontend to test that the implementation of tokens actually worked. I have renamed it to be clearer, and it can be used for validation of tokens when trying to access pages other than login

Reply:
I'm still not sure I understand.
Is this an example of an endpoint that requires authentication?
If so, that seems very verbose. Looking at the docs (https://fastapi.tiangolo.com/tutorial/security/oauth2-jwt/) it seems to me that they only need to add `current_user: Annotated[User, Depends(get_current_active_user)]` parameter and the endpoint is authenticated. Which is what I'd expect.

Reply:
I'll take a look at that, but I think this all might be better off in a branch for the "Authenticate BE Endpoints" task? In that case, I can remove it from here.

Reply:
Agreed